### ISSUE TITLE
Public Static Readonly Fields Trigger SA1401

### ISSUE THREAD

**Issue by lumberjack4** [DotNetAnalyzers/StyleCopAnalyzers#1320]
```
public static readonly MyClass Empty = new MyClass();
```

SA1401 is triggered telling me that the field must be private.  Recommend ignoring the rule for public static readonly fields, as this is the .NET paradigm for creating empty objects.

**Comment by sharwell** [comment]
Considering that C# 6 made it *extremely* easy to expose these as properties, my vote is that we not change the current behavior. The framework contains as many examples of properties being used for this purpose (e.g. `Encoding.UTF8`). To resolve the issue, you can actually *shorten* your code to the following, and improve your exposed API at the same time:

```csharp
public static MyClass Empty { get; } = new MyClass();
```

I'm leaving this marked **needs discussion** for now to get some more feedback.

**Comment by lumberjack4** [comment]
Cool!  I'm still trying to remember all the new changes.

**Comment by 0x084E** [comment]
See #1308

**Comment by bezysoftware** [comment]
What about dependency properties? Typically when you use the "propdp" snippet you get this generated:

private static readonly DependencyProperty FooProperty = DependencyProperty.Register("Foo", typeof(string), typeof(Class1), new PropertyMetadata(null));

which triggers SA1401.

**Comment by sharwell** [comment]
@oatkins Can you check the behavior of StyleCop "classic" with code like the previous comment?

**Comment by oatkins** [comment]
None of the following raises SA1401:

    public static readonly DependencyProperty MyPropertyProperty =
        DependencyProperty.Register("MyProperty", typeof(int), typeof(HasDependencyProperty), new PropertyMetadata(0));

    public static readonly string MyString = "fish";

    public static readonly StringBuilder Builder = new StringBuilder();

Interestingly, SC is being cautious here. My last example exposes a //mutable// object as a static field, which isn't good practice.

**Comment by vweijsters** [comment]
I would prefer that SA1401 will not be reported for `public static readonly` fields. The main reason being that we don't want to force people to change (a lot of) their code when migrating to StyleCop.Analyzers. A secondary reason to change the current behavior would be the dependency properties, which would be a major case for quite a few people I imagine.

**Comment by sharwell** [comment]
Moving this to a bug.
